src/sdk/pynni/nni/multi_phase/multi_phase_dispatcher.py

```diff
@@ -0,0 +1,177 @@
+# NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
+# OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# ==================================================================================================
+
+import logging
+from collections import defaultdict
+import json_tricks
+
+from nni.protocol import CommandType, send
+from nni.msg_dispatcher_base import MsgDispatcherBase
+from nni.assessor import AssessResult
+
+_logger = logging.getLogger(__name__)
+
+# Assessor global variables
+_trial_history = defaultdict(dict)
+'''key: trial job ID; value: intermediate results, mapping from sequence number to data'''
+
+_ended_trials = set()
+'''trial_job_id of all ended trials.
+We need this because NNI manager may send metrics after reporting a trial ended.
+TODO: move this logic to NNI manager
+'''
+
+def _sort_history(history):
+    ret = [ ]
+    for i, _ in enumerate(history):
+        if i in history:
+            ret.append(history[i])
+        else:
+            break
+    return ret
+
+# Tuner global variables
+_next_parameter_id = 0
+_trial_params = {}
+'''key: trial job ID; value: parameters'''
+_customized_parameter_ids = set()
+
+def _create_parameter_id():
+    global _next_parameter_id  # pylint: disable=global-statement
+    _next_parameter_id += 1
+    return _next_parameter_id - 1
+
+def _pack_parameter(parameter_id, params, customized=False, trial_job_id=None, parameter_index=None):
+    _trial_params[parameter_id] = params
+    ret = {
+        'parameter_id': parameter_id,
+        'parameter_source': 'customized' if customized else 'algorithm',
+        'parameters': params
+    }
+    if trial_job_id is not None:
+        ret['trial_job_id'] = trial_job_id
+    if parameter_index is not None:
+        ret['parameter_index'] = parameter_index
+    else:
+        ret['parameter_index'] = 0
+    return json_tricks.dumps(ret)
+
+class MultiPhaseMsgDispatcher(MsgDispatcherBase):
+    def __init__(self, tuner, assessor=None):
+        super()
+        self.tuner = tuner
+        self.assessor = assessor
+        if assessor is None:
+            _logger.debug('Assessor is not configured')
+
+    def load_checkpoint(self):
+        self.tuner.load_checkpoint()
+        if self.assessor is not None:
+            self.assessor.load_checkpoint()
+
+    def save_checkpoint(self):
+        self.tuner.save_checkpoint()
+        if self.assessor is not None:
+            self.assessor.save_checkpoint()
+
+    def handle_request_trial_jobs(self, data):
+        # data: number or trial jobs
+        ids = [_create_parameter_id() for _ in range(data)]
+        for i, _ in enumerate(ids):
+            params = self.tuner.generate_parameters(ids[i])
```

Comment:
I suggest to use `generate_multiple_parameters` here, because for some tuner algorithms calling `generate_parameters` one by one without feeding more data returns the same configuration.

Reply:
make sense, I will update it.